js/ui/windowManager.js

```diff
@@ -2166,3 +1888,368 @@ const WindowManager = new Lang.Class({
+                transition: ICON_BOUNCE_ANIMATION_TYPE_2,
+                delay: ICON_BOUNCE_ANIMATION_TIME * 0.25 + 0.3,
+                onComplete: function() {animateBounce(actor);}
+        });
+    }
+}
+
+const CodingManager = new Lang.Class({
+    Name: 'CodingManager',
+
+    _init: function(actor) {
+        this._sessions = [];
+        this._rotateInActors = [];
+        this._rotateOutActors = [];
+        this._firstFrameConnections = [];
+    },
+
+    addSwitcherToBuilder: function(actorApp) {
+        let window = actorApp.get_meta_window();
+
+        let button = new St.Button({ style_class: 'view-source' });
+        let rect = window.get_frame_rect();
+        button.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        Main.layoutManager.addChrome(button);
+
+        this._sessions.push({buttonApp: button,
+                             windowApp: actorApp,
+                             state: SessionState.APP});
+
+        let session = this._getSession(actorApp);
+        button.connect('clicked', Lang.bind(this, this._switchToBuilder, session));
+        let positionChangedId = window.connect('position-changed', Lang.bind(this, this._windowAppPositionChanged, session));
+        let windowsRestackedId = Main.overview.connect('windows-restacked', Lang.bind(this, this._windowAppRestacked, session));
+        session.positionChangedIdApp = positionChangedId;
+        session.windowsRestackedIdApp = windowsRestackedId;
+    },
+
+    addSwitcherToApp: function(actorBuilder, windowApp) {
+        let currentSession = this._sessions.filter(function(session) {
+            return (session.windowApp.get_meta_window() === windowApp);
+        });
+        if (currentSession.length === 0)
+            return;
+
+        let session = currentSession[0];
+        session.windowBuilder = actorBuilder;
+
+        this._animateToBuilder(session);
+    },
+
+    removeSwitcherToBuilder : function(actorApp) {
+        let session = this._getSession(actorApp);
+
+        if (session.positionChangedIdApp != 0) {
+            session.windowApp.get_meta_window().disconnect(session.positionChangedIdApp);
+            session.positionChangedIdApp = 0;
+        }
+        if (session.windowsRestackedIdApp != 0) {
+            Main.overview.disconnect(session.windowsRestackedIdApp);
+            session.windowsRestackedIdApp = 0;
+        }
+        session.buttonApp.destroy();
+        if (!this._removeSession(session))
+            return false;
+        return true;
+    },
+
+    removeSwitcherToApp : function(actorBuilder) {
+        let session = this._getSession(actorBuilder);
+
+        if (session.positionChangedIdBuilder != 0) {
+            session.windowBuilder.get_meta_window().disconnect(session.positionChangedIdBuilder);
+            session.positionChangedIdBuilder = 0;
+        }
+        session.buttonBuilder.destroy();
+        session.windowBuilder = null;
```

Comment:
It may still be better to give the buttons their own class. I don't think this class should have to know about managing their details.